firebase-functions/functions/src/choiceUpdated.ts

```diff
@@ -1,29 +1,34 @@
 ﻿import * as functions from 'firebase-functions';
 import * as admin from 'firebase-admin';
+import { calculateLosers, hasEveryoneChosen } from './choiceOperations';
 
-// Dummy implementation for now. Will set some chosen thing just something will be visible
-export const choiceUpdated = functions.firestore.document('/players/{documentId}')
+export default functions.firestore.document('/players/{documentId}')
   .onUpdate(async (change: any, context: any) => {
-    const playerId = context.params.documentId;
     try {
-      const doc = await admin.firestore().collection('players').doc(playerId).get();
-      const roomId = doc.data()?.room;
-      console.log('The room in question ' + roomId);
-      const snapshot = await admin.firestore().collection('players').where('room', '==', roomId).get();
-      let areAllChoicesPresent = true;
-      let choice = '';
-      snapshot.forEach((player: any) => {
-        if (!player.data().choice) {
-          areAllChoicesPresent = false;
-          return;
-        }
-        choice = player.data().choice;
-      });
-
-      if (areAllChoicesPresent) {
-        await admin.firestore().collection('rooms').doc(roomId).update({ winner: choice });
+      const roomId = await getCurrentRoomId(context);
+      const playersInSameRoom = (await admin.firestore().collection('players').where('room', '==', roomId).get()).docs;
+      if (!hasEveryoneChosen(playersInSameRoom)) {
+        return;
+      }
+      const losers = calculateLosers(playersInSameRoom);
+      if (!losers.length) {
+        await Promise.all(resetAllChoices(playersInSameRoom));
+        return;
       }
+      // deactivate all the losers and disable them in the FE (maybe even check if all the choices were from active users first)
+      // if only one player remains, set him as the winner and display in FE
+      await admin.firestore().collection('rooms').doc(roomId).update({ winner: losers[0] });
     } catch (e) {
       console.log(e);
     }
   });
+
+async function getCurrentRoomId(context: any): Promise<string> {
```

Comment:
maybe pass the `documentId` as type of `string `as argument

Reply:
Good idea, done